www/js/splash/notifScheduler.js

```diff
@@ -0,0 +1,190 @@
+
+    const scheduler = {};
+    let _config;
+    let scheduledPromise = new Promise((rs) => rs());
+    let isScheduling = false;
+
+    // like python range()
+    function range(start, stop, step) {
+        let a = [start], b = start;
+        while (b < stop)
+            a.push(b += step || 1);
+        return a;
+    }
+
+    // returns an array of moment objects, for all times that notifications should be sent
+    const calcNotifTimes = (scheme, dayZeroDate, timeOfDay) => {
+        const notifTimes = [];
+        for (const s of scheme.schedule) {
+            // the days to send notifications, as integers, relative to day zero
+            const notifDays = range(s.start, s.end, s.intervalInDays);
+            for (const d of notifDays) {
+                const date = moment(dayZeroDate).add(d, 'days').format('YYYY-MM-DD')
+                const notifTime = moment(date+' '+timeOfDay, 'YYYY-MM-DD HH:mm');
+                notifTimes.push(notifTime);
+            }
+        }
+        return notifTimes;
+    }
+
+    // returns true if all expected times are already scheduled
+    const areAlreadyScheduled = (notifs, expectedTimes) => {
+        for (const t of expectedTimes) {
+            if (!notifs.some((n) => moment(n.at).isSame(t))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    const setUpActions = () => {
+        const action = {
+            id: 'action',
+            title: 'Change Time',
+            launch: true
+        };
+        return new Promise((rs) => {
+            cordova.plugins.notification.local.addActions('reminder-actions', [action], rs);
+        });
+    }
+
+    function debugGetScheduled(prefix) {
+        cordova.plugins.notification.local.getScheduled((notifs) => {
+            if (!notifs?.length)
+                return Logger.log(`${prefix}, there are no scheduled notifications`);
+            const time = moment(notifs?.[0].trigger.at).format('HH:mm');
+            Logger.log(`${prefix}, there are ${notifs.length} scheduled notifications at ${time}`);
+        });
+    }
+
+    // schedules the notifications using the cordova plugin
+    const scheduleNotifs = (scheme, notifTimes) => {
+        return new Promise((rs) => {
+            isScheduling = true;
+            const nots = notifTimes.map((n) => {
+                const nDate = n.toDate();
+                const seconds = nDate.getTime() / 1000;
+                return {
+                    id: seconds,
+                    title: scheme.title,
+                    text: scheme.text,
+                    trigger: {at: nDate},
+                    actions: 'reminder-actions',
+                    data: {
+                        action: {
+                            redirectTo: 'root.main.control',
+                            redirectParams: {
+                                openTimeOfDayPicker: true
+                            }
```

Comment:
The existing mechanism only handled redirect paths when the notification itself is clicked, but the desired redirect in this instance happens when the "Change Time" button is clicked. So I needed a way to specify the redirect that should happen in the 'action' state.
`localNotify.getRedirectState()` was tweaked so that we can re-use the existing logic in both instances, depending on if `data` is passed as a parameter or if `data.action` is passed

Reply:
Ok, I get it now.

My original thought was that, since we don't have a redirect for the default click, we should be able to specify this redirect there. But repurposing that way means that whenever the user clicks on the notification, they will be redirected to the profile instead of opening the app at the label screen. Whoops!

So we do need this to indicate that we only want to redirect when the action is clicked.